### PR TITLE
refact(getter): refactor the KS Cloud client

### PR DESCRIPTION
**Pull request by fredbi**
This PRs proposes a significant revamping of the KSCloud API HTTP client.

* Interfaces are unchanged

* Deprecated: low-level API funcs marked for deprecation: HttpPost, HttpGetter, HttpDelete (an augmented version of the KS Cloud client will expose the post report API, which is currently the sole use-case of low-level API)

* Doc: the package is now godoc-friendly

* Style & code layout:
  * listed all exposed types via aliases, for clarity/less confusing imports
  * unexported private types
  * factorized query param logic
  * factorized type <-> JSON using generic func & io.Reader
  * "utils" are now limited to a few common utility functions
  * centralized hard-coded strings as (unexported) constants
  * concision: use higher-level http definitions such as constants, cookie methods, etc
  * included type-safety guards to verify that interfaces are actually implemented by the exported types

* Tests: existing test assertions are unchanged
  * tests are beefed-up to assert proper authentication flow (token & cookie).
  * added unit tests for utility methods

* Perf:
  * unmarshalling API responses is now flowing without extraneous memory allocation via string representation
  * request headers are now passed withot extraneous map allocation
  * JSON operations are now fully supported by jsoniter (no longer use encoding/json)

* Changes in functionality:
  * the client is now fully extensible with KSCloudOption
  * use the option functor idiom to keep constructors short
  * methods that used to mute errors (i.e. return nil, nil) now bubble up errors
  * the captured cookie is now captured in full, not just its value (other cookie parameters returned are stored)
  * added a request/response dump option, for debugging

### CODE REVIEW

**Comment by fredbi** [comment]
cc @dwertent @matthyx  a proposal for a cleaner, more extensible HTTP API client (now that the tests are merged)

**Comment by fredbi** [comment]
@matthyx @dwertent 
This change aims to centralize all dealings with the SaaS API in one place, thus paving the way for a simplification of other packages (to wit, reports uploading).

I believe it allows us to better cope with future changes of the API.

**Review by matthyx — APPROVED** [review]
Never ever such a big PR... took me 2 days to review and I would like to perform some manual tests with our platform before merging.
But great job, I really like the cleaned interfaces !

**Comment by matthyx** [comment]
@fredbi we'll merge this after Sunday's hotfix release

File: core/cautils/getter/getpoliciesutils_test.go
Comment (fredbi) [reply core/cautils/getter/getpoliciesutils_test.go:44]: You’re right. This should be just “json”

Comment (fredbi) [reply core/cautils/getter/getpoliciesutils_test.go:44]: Not really impacting however.

Comment (fredbi) [reply core/cautils/getter/getpoliciesutils_test.go]: @matthyx fixed this omission of mine. Thank you.

**Comment by fredbi** [comment]
@matthyx
 
> Never ever such a big PR... took me 2 days to review and I would like to perform some manual tests with our platform before merging.

Ok. Copy that. Splitting other outstanding PRs in small parts now. My apologies for having gone overboard there.

File: core/cautils/getter/kscloudoptions.go
Comment (dwertent) [inline core/cautils/getter/kscloudoptions.go:180]: [TODO]: Use the [Kubescape logger](https://github.com/kubescape/go-logger) instead

File: core/cautils/getter/kscloudoptions.go
Comment (dwertent) [inline core/cautils/getter/kscloudoptions.go:190]: [TODO]: Use the [Kubescape logger](https://github.com/kubescape/go-logger) instead

**Review by dwertent — APPROVED** [review]
@fredbi  Well done!

File: core/cautils/getter/kscloudoptions.go
Comment (fredbi) [reply core/cautils/getter/kscloudoptions.go:180]: @dwertent ok well noted. Will do on the next round.

**Comment by fredbi** [comment]
@matthyx I've address your legit concern about jsoniter imports remaining here and there.
TL,DR: I just had to upgrade my vim plugin - the automatic goimport had some flaw with go1.19+...

File: core/cautils/getter/kscloudoptions.go
Comment (fredbi) [reply core/cautils/getter/kscloudoptions.go:180]: @dwertent  there are also a few adjustments to be carried out with the utm query params to be done (realized that when finishing up this scaffolding in #1134). I'll issue a small follow-up PR this week to separate these different concerns and spare a tedious PR review. 

Thanks again all for your trust and super-supportive attitude. I hope that more activity and efforts on code quality will help this raw diamond to gain a lot of traction among the dev community. That would be my humble contribution...

**Review by matthyx — APPROVED** [review]
@dwertent let's merge this one as well

**Pull request merged**
